apps/vmq_diversity/src/vmq_diversity_worker_wrapper.erl

```diff
@@ -0,0 +1,191 @@
+%% Copyright 2017 Erlio GmbH Basel Switzerland (http://erl.io)
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+
+%% @doc Simple gen_server for poolboy used to wrap a client process
+%% and restart (after a configurable timeout) it if it dies for some
+%% reason. The purpose is to protect poolboy from dying if an external
+%% resource is not available and all the client processes die.
+%% @end.
+-module(vmq_diversity_worker_wrapper).
+
+-behaviour(gen_server).
+
+%% API
+-export([start_link/1, apply/4]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2, code_change/3]).
+
+-define(SERVER, ?MODULE).
+
+-record(state, {
+          %% module, fun, args used to start the worker.
+          start_mfa = undefined,
```

Comment:
why so explicit?

Reply:
My idea was to be general and not make any assumptions about the names and arguments.

Reply:
I guess it could just be a fun of arity 0 being passed, so we only store a reference to the function, and we wouldn't have to use apply at all for starting the wrapped process. Think that's cleaner.